src/NuGet.Core/NuGet.Protocol/Resources/ServiceIndexResourceV3.cs

```diff
@@ -119,6 +136,16 @@ public virtual Uri GetServiceEntryUri(params string[] orderedTypes)
             return GetServiceEntryUris(clientVersion, orderedTypes).FirstOrDefault();
         }
 
+        /// <summary>
+        /// Get the best match service URI. Give the option to allow insecure connections or non https sources.
+        /// Default is false.
+        /// </summary>
+        public virtual Uri GetServiceEntryUri(bool allowInsecureConnections, params string[] orderedTypes)
+        {
+            _allowInsecureConnections = allowInsecureConnections;
```

Comment:
This is not "thread safe", and I believe that currently `SourceRepository` and therefore `ServiceIndexResourceV3` can safely be called on multiple threads at the same time. In fact, I'm pretty sure the design of our VS integration actively uses it.

Also, from a public API design point of view, I wouldn't allow callers to specify whether they want to allow insecure connections. "Normal" SourceRepository usage is to create a SourceRepository object from a PackageSource, and then all "resources" use the settings from that initial creation. Therefore, I think it would be better that when this `ServiceIndexResourveV3` object is created, the `allowInsecureConnections` bool is passed from the `packageSource`, and then you don't need new public `GetServiceEntryUri` methods with the parameter.